server/app/interfaces: update AAS- and SubmodelRepository API to V3.1

Frosty2500 opened the pull request:
This PR updates the AAS- and SubmodelRepository servers from version 3.01 to version 3.1 and fixes some bugs.

Frosty2500 committed:
update AAS_Repository to V3.1

Frosty2500 committed:
fix pycodestyle errors

Frosty2500 committed:
update docstring

Frosty2500 committed:
add deterministic pagination to submodel_refs

s-heppner commented:
We need to wait with merging this PR until we release v2.0.0.

Frosty2500 committed:
update README, API is version 3.1.1

Frosty2500 committed:
remove cursor when end of result set reached

s-heppner committed:
Merge branch 'develop' into update/api_V3.1

s-heppner commented:
I've updated this branch, the merge conflicts were quite complex due to the refactoring we did while this PR was already out. 

Currently, there's a typing issue:

```Python
    @classmethod
    def _get_slice(cls, request: Request, iterator: Iterable[T]) -> Tuple[Iterator[T], int]:
        limit_str = request.args.get('limit', default="10")
        cursor_str = request.args.get('cursor', default="1")
        try:
            limit, cursor = int(limit_str), int(cursor_str) - 1  # cursor is 1-indexed
            if limit < 0 or cursor < 0:
                raise ValueError
        except ValueError:
            raise BadRequest("Limit can not be negative, cursor must be positive!")
        start_index = cursor
        end_index = cursor + limit
        items = list(itertools.islice(iterator, start_index, end_index + 1))
        has_more = len(items) > limit
        paginated_slice = iter(items[:limit])
        next_cursor = cursor + limit if has_more else None
        return paginated_slice, next_cursor
```

As it is implemented, it actually returns `Optional[int]` with `next_cursor`. 
However, the endpoints that use this method do expect `int`. 
Someone more experienced with the API spec should investigate how this method is supposed to behave and what we need to adapt. 

Furthermore, @moritzsommer, could you review the `repository.py`  file w.r.t. the `ObjectStore` refactor please?

s-heppner commented:
I did some research and found:

> Constraint AASa-001: The value of the [cursor](https://industrialdigitaltwin.io/aas-specifications/IDTA-01002/v3.1.1/http-rest-api/http-rest-api.html#pagination) query parameter must not be empty. If the client does not know the cursor value, it must omit the whole query parameter in the request.

But this is with regards to how the client needs to form the request, not how the server needs to respond.

hpoeche commented:
<!-- obsidian --><p>The pagination of results as defined in AAS Part 2, <a href="https://industrialdigitaltwin.io/aas-specifications/IDTA-01002/v3.1.2/http-rest-api/http-rest-api.html#pagination" class="external-link" target="_blank" rel="noopener nofollow" aria-label="https://industrialdigitaltwin.io/aas-specifications/IDTA-01002/v3.1.2/http-rest-api/http-rest-api.html#pagination" data-tooltip-position="top">"Pagination"</a> uses the <code>cursor</code> parameter in requests for defining the position from which to resume the results in the next request. The required value is obtained from a previous paginated result, containing a <code>paging_metadata</code> which is defined in the specs as follows:</p>

Attribute | Explanation | Type | Cardinality
-- | -- | -- | --
cursor | The cursor for the next part of the result set. No cursor attribute means that the end of the result set has been reached. | string | 0..1


<p>In the server implementation all logic in handling the <code>cursor</code> value is encapsulated in the  <code>_get_slice(...)</code> method. Every endpoint implementation that invokes this method at some point, passes the value of <code>next_cursor</code> unmodified to the <code>cursor</code> parameter in the <code>APIResponse</code> constructor, which is typed <code>Optional[int]</code>.</p>
<p>Therefore I suggest to adapt the return type of <code>_get_slice(...)</code> (as well as all other methods that depend on it) to <code>Tuple[Iterator[T], Optional[int]]</code>.</p>
<p>Despite that, I want to point out that the current implementation of <code>_get_slice(...)</code> is incorrect, as the returned <code>next_cursor</code> value is too low by 1 (due to the 1-based indexing). Additionally, the AAS Spec Part 2, defines the default value of the <code>limit</code> request parameter to be <code>100</code> rather than <code>10</code>.</p>

s-heppner committed:
Make `cursor` optional everywhere

Previously, the `server.app.interfaces.base._get_slice()` method was
adapted to return an `Optional[int]` cursor. However, the useages of
this method were not adapted, leading to failing static analysis.

This adapts the type hints of the usages of the `_get_slice()` method.

s-heppner committed:
Merge branch 'develop' into update/api_V3.1

s-heppner committed:
Fix copyright notice

s-heppner committed:
Merge branch 'develop' into update/api_V3.1

s-heppner commented on server/README.md:
```diff
@@ -136,9 +136,9 @@ This Dockerfile is inspired by the [tiangolo/uwsgi-nginx-docker][10] repository.
-[6]: https://industrialdigitaltwin.io/aas-specifications/IDTA-01002/v3.0/index.html
+[4]: https://app.swaggerhub.com/apis/Plattform_i40/AssetAdministrationShellRepositoryServiceSpecification/V3.1.1_SSP-001
+[5]: https://app.swaggerhub.com/apis/Plattform_i40/SubmodelRepositoryServiceSpecification/V3.1.1_SSP-001
+[6]: https://industrialdigitaltwin.org/content-hub/aasspecifications/idta_01002-3-0_application_programming_interfaces
```
```suggestion
[6]: https://industrialdigitaltwin.io/aas-specifications/IDTA-01002/v3.1.2/index.html
```

s-heppner committed:
Apply suggestion from @s-heppner
